department-of-veterans-affairs/vets-api
# 82171 - Implement Sign-in Service configuration certificate checks

## Summary

This PR implements daily checks for problematic certificates in client and service account configurations. 

I extracted shared certificate-related code from the `SignIn::ServiceAccountConfig` and `SignIn::ClientConfig` models and moved it into a model concern.

I implemented additional logic for selecting expired, expiring, or self-signed certificates in the concern.

I then implemented a daily job that sends logs to Datadog for each expired, expiring or self-signed certificate.

## Related issue(s)

https://app.zenhub.com/workspaces/identity-5f5bab705a94c9001ba33734/issues/gh/department-of-veterans-affairs/va.gov-team/82171

## Testing done

- Fire up a Rails console.
```
bin/rails c
```
- Add some invalid certificates to a configuration.
```ruby
config = SignIn::ServiceAccountConfig.last

expired_certificate = File.read('spec/fixtures/sign_in/sample_expired_client.crt')
config.certificates << expired_certificate

self_signed_certificate = File.read('spec/fixtures/sign_in/sample_self_signed_client.crt')
config.certificates << self_signed_certificate

config.save
```
- Kick off the `SignIn::CertificateCheckerJob`
```ruby
SignIn::CertificateCheckerJob.new.perform
```
- Verify each invalid certificate has a corresponding log entry with the config and certificate details. Example:
```
2024-05-15 10:52:49.053702 I [76600:12160 certificate_checker_job.rb:30] Rails -- [SignIn] [CertificateChecker] self_signed_certificate -- { :config_type => "SignIn::ServiceAccountConfig", :config_id => 1, :config_description => "VA Identity Dashboard API", :certificate_subject => "/C=US/ST=Oregon/L=Portland/O=Oddball", :certificate_issuer => "/C=US/ST=Oregon/L=Portland/O=Oddball", :certificate_serial => "11378111649178017353", :certificate_not_before => "2023-07-20 13:36:11 UTC", :certificate_not_after => "2024-07-19 13:36:11 UTC" }
```


## Screenshots

Not relevant.

## What areas of the site does it impact?

Sign-in Service.

## Acceptance criteria
